pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java

```diff
@@ -84,7 +84,18 @@ public void initialize(ServiceConfiguration conf, PulsarResources pulsarResource
     @Override
     public CompletableFuture<Boolean> canProduceAsync(TopicName topicName, String role,
             AuthenticationDataSource authenticationData) {
-        return checkAuthorization(topicName, role, AuthAction.produce);
+        return validateTenantAdminAccess(topicName.getTenant(), role, authenticationData)
+                .thenCompose(isSuperUserOrAdmin -> {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Verify if role {} is allowed to produce topic {}: isSuperUserOrAdmin={}",
+                                role, topicName, isSuperUserOrAdmin);
+                    }
+                    if (isSuperUserOrAdmin) {
+                        return CompletableFuture.completedFuture(true);
+                    } else {
+                        return checkAuthorization(topicName, role, AuthAction.produce);
+                    }
+                });
```

Comment:
I don't think we want to modify the `canProduceAsync` method. I think we should migrate the calling code to use `allowTopicOperationAsync` since that internally calls `validateTenantAdminAccess` then calls `canProduceAsync`.

Reply:
@dragonls - I just noticed the rest of the conversation above. What I don't see is an answer explaining why we don't migrate from `canProduceAsync` to `allowTopicOperationAsync`. Do you mind addressing that point? Thanks!

Reply:
> 

I think you are saying migrate from `org.apache.pulsar.broker.authorization.AuthorizationProvider#canProduceAsync` to `org.apache.pulsar.broker.authorization.AuthorizationProvider#allowTopicOperationAsync` in `org.apache.pulsar.broker.authorization.AuthorizationService#canProduceAsync`. If do so, this bug is indeed fixed.

However, this still leave hidden dangers in `org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider`. The execution effect of `org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider#allowTopicOperationAsync` and `org.apache.pulsar.broker.authorization.PulsarAuthorizationProvider#canProduceAsync` is still not consistent, because the latter do not check tenant admin permission.

Reply:
In AuthenticationProvider, the `allowTopicOperationAsync` calls the `canProduceAsync`, and the `canConsumeAsync`, when we use the `allowTopicOperationAsync`, it works fine.

However, in the AuthenticationService, some callers call the `canProduceAsync`, the `canConsumeAsync`, and the `allowTopicOperationAsync`, when calling the `canProduceAsync`, we encountered this bug.

So I made https://github.com/apache/pulsar/pull/19461.

Reply:
I might understand what @michaeljmarshall meant.

We can make `canProduceAsync` and `canConsumeAsync` to call `allowTopicOperationAsync` instead. And `allowTopicOperationAsync` calls maybe `innerCanProduceAsync` and `innerCanConsumeAsync`, which do the same thing as the old `canProduceAsync` do.

Reply:
> I think you are saying migrate from `org.apache.pulsar.broker.authorization.AuthorizationProvider#canProduceAsync` to `org.apache.pulsar.broker.authorization.AuthorizationProvider#allowTopicOperationAsync` in `org.apache.pulsar.broker.authorization.AuthorizationService#canProduceAsync`. If do so, this bug is indeed fixed.

Yes, that was my point. Instead of changing the meaning of `AuthorizationService#canProduceAsync`, I think we should change which method we call.

Reply:
I think it is risky to change the meaning of the underlying method. I will review https://github.com/apache/pulsar/pull/19461 more closely, but I agree with the general premise that the `AuthorizationService` is not correctly using the `AuthenticationProvider` framework.

Reply:
I agreed this idea.

Reply:
> > I think you are saying migrate from `org.apache.pulsar.broker.authorization.AuthorizationProvider#canProduceAsync` to `org.apache.pulsar.broker.authorization.AuthorizationProvider#allowTopicOperationAsync` in `org.apache.pulsar.broker.authorization.AuthorizationService#canProduceAsync`. If do so, this bug is indeed fixed.
> 
> Yes, that was my point. Instead of changing the meaning of `AuthorizationService#canProduceAsync`, I think we should change which method we call.

Updated